File: Java/TollCalculator.java

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.temporal.ChronoField;
import java.time.temporal.TemporalField;
import java.util.*;
import java.util.concurrent.*;

public class TollCalculator {

  /**
   * Calculate the total toll fee for one day
   *
   * @param vehicle - the vehicle
   * @param dates   - date and time of all passes on one day
   * @return - the total toll fee for that day
   */
  public int getTollFee(Vehicle vehicle, Date... dates) {
    // I believe in this method the times 08:40 and 09:20 should consider as times of 2 different hours (i.e. hour of 8 and hour of 9.
    // In the previous logic if the 08:40 was the first of the array then 09:20 was considered as same hour because it has considered the time gap
    // Also the previous implementation was not updating the intervalStart correctly.

    int[] feeOfTheHour = new int[24];
    Arrays.fill(feeOfTheHour, 0);

    // calculate max fee for each hour
    for (Date date : dates) {
      int fee = getTollFee(date, vehicle);
      int hour = Instant.ofEpochMilli(date.getTime()).atZone(ZoneId.systemDefault()).getHour();

      if(feeOfTheHour[hour] < fee){
        feeOfTheHour[hour] = fee;
      }
    }

    // get the sum of hourly fee
    int totalFee = Arrays.stream(feeOfTheHour).sum();

    // cap the value
    if (totalFee > 60) totalFee = 60;

    return totalFee;
  }

  private boolean isTollFreeVehicle(Vehicle vehicle) {
    if(vehicle == null) return false;
    String vehicleType = vehicle.getType();
    return vehicleType.equals(TollFreeVehicles.MOTORBIKE.getType()) ||
           vehicleType.equals(TollFreeVehicles.TRACTOR.getType()) ||
           vehicleType.equals(TollFreeVehicles.EMERGENCY.getType()) ||
           vehicleType.equals(TollFreeVehicles.DIPLOMAT.getType()) ||
           vehicleType.equals(TollFreeVehicles.FOREIGN.getType()) ||
           vehicleType.equals(TollFreeVehicles.MILITARY.getType());
  }

  public int getTollFee(final Date date, Vehicle vehicle) {
    if(isTollFreeDate(date) || isTollFreeVehicle(vehicle)) return 0;
    Calendar calendar = GregorianCalendar.getInstance();
    calendar.setTime(date);
    int hour = calendar.get(Calendar.HOUR_OF_DAY);
    int minute = calendar.get(Calendar.MINUTE);

    if (hour == 6 && minute >= 0 && minute <= 29) return 8;
    else if (hour == 6 && minute >= 30 && minute <= 59) return 13;
    else if (hour == 7 && minute >= 0 && minute <= 59) return 18;
    else if (hour == 8 && minute >= 0 && minute <= 29) return 13;
    else if (hour == 8 && minute >= 30 && minute <= 59) return 8;
    else if (hour >= 9 && hour <= 14 && minute >= 0 && minute <= 59) return 8;
    else if (hour == 15 && minute >= 0 && minute <= 29) return 13;
    else if (hour == 15 && minute >= 30 && minute <= 59) return 18;
    else if (hour == 16 && minute >= 0 && minute <= 59) return 18;
    else if (hour == 17 && minute >= 0 && minute <= 59) return 13;
    else if (hour == 18 && minute >= 0 && minute <= 29) return 8;
    else return 0;
  }

  private Boolean isTollFreeDate(Date date) {
    Calendar calendar = GregorianCalendar.getInstance();
    calendar.setTime(date);
    int year = calendar.get(Calendar.YEAR);
    int month = calendar.get(Calendar.MONTH);
    int day = calendar.get(Calendar.DAY_OF_MONTH);

    int dayOfWeek = calendar.get(Calendar.DAY_OF_WEEK);
    if (dayOfWeek == Calendar.SATURDAY || dayOfWeek == Calendar.SUNDAY) return true;

    //TODO - consider every year differently or remove the year validation
    if (year == 2013) {
      if (month == Calendar.JANUARY && day == 1 ||
          month == Calendar.MARCH && (day == 28 || day == 29) ||
          month == Calendar.APRIL && (day == 1 || day == 30) ||
          month == Calendar.MAY && (day == 1 || day == 8 || day == 9) ||
          month == Calendar.JUNE && (day == 5 || day == 6 || day == 21) ||
          month == Calendar.JULY ||
          month == Calendar.NOVEMBER && day == 1 ||
          month == Calendar.DECEMBER && (day == 24 || day == 25 || day == 26 || day == 31)) {
        return true;
      }
    }
    return false;
  }

  private enum TollFreeVehicles {
    MOTORBIKE("Motorbike"),
    TRACTOR("Tractor"),
    EMERGENCY("Emergency"),
    DIPLOMAT("Diplomat"),
    FOREIGN("Foreign"),
    MILITARY("Military");
    private final String type;

    TollFreeVehicles(String type) {
      this.type = type;
    }

    public String getType() {
      return type;
    }
  }
}

File: PHP/src/TollCalculator.php
<?php

namespace App;

use DateTime;

class TollCalculator {

  public const NO_FEE = 0;

  public const LOW_FEE = 8;

  public const MEDIUM_FEE = 13;

  public const HIGH_FEE = 18;

  public const MAX_FEE = 60;

  /**
   * An array of toll free dates.
   *
   * @var array
   * @todo This should ideally not be hard-coded but rather fetch from an
   *   API. That is however out of scope.
   *
   */
  protected array $tollFreeDates = [
    '2021' => [
      '1' => ['1', '5', '6'],
      '4' => ['1', '2', '5', '30'],
      '5' => ['12', '13'],
      '6' => ['25'],
      '11' => ['5'],
      '12' => ['24', '31'],
    ],
  ];

  /**
   * Calculate the total toll fee for one day.
   *
   * @param \App\VehicleInterface $vehicle
   * @param \DateTime[] $date_times
   *
   * @return int
   *   Fee in SEK.
   */
  public function calculateTollFee(VehicleInterface $vehicle, array $date_times): int {
    if ($vehicle->isTollFree()) {
      return self::NO_FEE;
    }

    if ($this->isTollFreeDate($date_times[0])) {
      return self::NO_FEE;
    }

    $total_fee = 0;

    $interval_start = $date_times[0];
    foreach ($date_times as $index => $date_time) {
      $fee = $this->getTollFee($date_time);

      $diff_hours = $date_time->diff($interval_start)->h;

      if ($index !== 0 && $diff_hours < 1) {
        $previous_fee = $this->getTollFee($interval_start);

        if ($fee <= $previous_fee) {
          continue;
        }

        $total_fee -= $previous_fee;
      }

      if ($diff_hours >= 1) {
        $interval_start = $date_time;
      }

      $total_fee += $fee;
    }

    if ($total_fee > self::MAX_FEE) {
      return self::MAX_FEE;
    }

    return $total_fee;
  }

  /**
   * Get toll fee for a single passage.
   *
   * @param \DateTime $date_time
   *
   * @return int
   */
  public function getTollFee(DateTime $date_time): int {
    $date_from_input = $date_time->format('Y-m-d');

    $fee_intervals = [
      ['00:00', '05:59', self::NO_FEE],
      ['06:00', '06:29', self::LOW_FEE],
      ['06:30', '06:59', self::MEDIUM_FEE],
      ['07:00', '07:59', self::HIGH_FEE],
      ['08:00', '08:29', self::MEDIUM_FEE],
      ['08:30', '14:59', self::LOW_FEE],
      ['15:00', '15:29', self::MEDIUM_FEE],
      ['15:30', '16:59', self::HIGH_FEE],
      ['17:00', '17:59', self::MEDIUM_FEE],
      ['18:00', '18:29', self::LOW_FEE],
      ['18:30', '23:59', self::NO_FEE],
    ];

    foreach ($fee_intervals as [$start_time, $end_time, $fee]) {
      $interval_start = new DateTime("$date_from_input $start_time");
      $interval_end = new DateTime("$date_from_input $end_time");

      if ($date_time >= $interval_start && $date_time <= $interval_end) {
        return $fee;
      }
    }

    return self::NO_FEE;
  }

  /**
   * Check whether the date of passage is a toll free date.
   *
   * Saturdays, Sundays and bank holidays are toll free.
   *
   * @param \DateTime $date
   *
   * @return bool
   */
  protected function isTollFreeDate(DateTime $date): bool {
    $year = $date->format('Y');
    $month = $date->format('n');
    $day = $date->format('j');
    $weekday = $date->format('l');

    if ($weekday === 'Saturday' || $weekday === 'Sunday') {
      return TRUE;
    }

    return isset($this->tollFreeDates[$year][$month][$day]);
  }

}
